File: python/117.py

# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root == None: return None
        q = [root]
        while q:
            for i in range(len(q) - 1):
                q[i].next = q[i+1]
            tmp = q
            q = []
            for node in tmp:
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)
        return root

File: go/117.go
package main

type Node1 struct {
	Val   int
	Left  *Node1
	Right *Node1
	Next  *Node1
}

func connect(root *Node1) *Node1 {
	if root == nil {
		return nil
	}
	nodeList := []*Node1{root}
	levelList := []int{0}
	i := 0
	var node *Node1
	for i < len(nodeList) {
		node = nodeList[i]
		if node.Left != nil {
			nodeList = append(nodeList, node.Left)
			levelList = append(levelList, levelList[i]+1)
		}
		if node.Right != nil {
			nodeList = append(nodeList, node.Right)
			levelList = append(levelList, levelList[i]+1)
		}
		i++
	}
	for i := 0; i < len(nodeList)-1; i++ {
		if levelList[i] == levelList[i+1] {
			nodeList[i].Next = nodeList[i+1]
		}
	}
	return root
}

// func main() {

// }
